### PR TITLE
PMMR Block to File positions metadata storage

### PR DESCRIPTION
**Pull request by yeastplume**
For https://github.com/mimblewimble/grin/issues/735

Will fill more detail here later, but the general approach is to collect all of the file positions when a block is commit to the dataset and the files are synced, then store this metadata within the DB by block hash. This should allow us to build a list of rollback positions for the PMMR files, however for a first out we'll just limit storage to the current block hash and the previous block.

### CODE REVIEW

**Comment by antiochp** [comment]
Question - do we need this for every block or just the last known good one? 
I'm not sure if we gain much by tracking this for every block (but I guess this would support reverting to any historical block).

File: chain/src/pipe.rs [hunk 1]
```diff
@@ -108,7 +108,49 @@ pub fn process_block(b: &Block, mut ctx: BlockContext) -> Result<Option<Tip>, Er
...
+		},
+		Err(e) => Err(e),
+	}
+}
+
+fn sync_file_metadata(b: &Block, sumtrees: &sumtree::SumTrees, store: Arc<ChainStore>) -> Result<(), Error> {
+	// We want to store written file data for the last 2 blocks, so we potentially
+	// have a couple of options to roll back to
+
+	// TODO: Delete all old blocks before doing this
```
Comment (ignopeverell) [inline chain/src/pipe.rs]: I think we can handle the deletion at the same time we do compaction. So we'd delete the old blocks, delete the metedata record, and compact the MMR regularly.

File: chain/src/pipe.rs [hunk 2]
```diff
@@ -108,7 +108,49 @@ pub fn process_block(b: &Block, mut ctx: BlockContext) -> Result<Option<Tip>, Er
...
+	// Delete the metadata of the oldest block written
+	let oldest_block_hash = store
+		.get_previous_pmmr_file_block();
+	if let Ok(h) = oldest_block_hash {
+		store
+			.delete_block_pmmr_file_metadata(&h)
+			.map_err(|e| Error::StoreErr(e, "deleting oldest stored pmmr block".to_owned()))?;
+	}
+
+	// Swap previous block index
```
Comment (ignopeverell) [inline chain/src/pipe.rs]: Instead of the 2 below wouldn't it be simpler (and more atomic) to save that directly in the chain head?

**Comment by yeastplume** [comment]
@antiochp Originally we'd talked about storing the last 'n' blocks, where n is a smallish number, so we'd have options to roll back to in the event of corruption, but I think the last block and the previous one to that is enough to start with. I've implemented the data storage to store file positions by hash so we can theoretically add more historical block positions if needed.

File: chain/src/pipe.rs (hunk 1 above)
Comment (yeastplume) [reply chain/src/pipe.rs]: Yes, so long as compaction happens fairly regularly.. just worried about leaving stale pairings around in the DB in the event of forks, and I don't think we actually have any store functions implemented yet to delete all records by their prefix (sure it's trivial, just haven't looked up the docs). Compaction's probably a good place to do this

File: chain/src/pipe.rs (hunk 2 above)
Comment (yeastplume) [reply chain/src/pipe.rs]: Absolutely, good point, I can just add a field to the Tip struct and save it then, then we just have chain head and previous block position to deal with for the time being.

**Comment by yeastplume** [comment]
updated now to just use the Tip fields for the current and previous blocks.. these should be okay in a fork, because block hash/PMMR pairings are just added every time a block is successfully processed therefore the paring should always be valid for the latest version of a block (until compaction at least, which should remove all of the pairings)

File: chain/src/chain.rs [hunk 3]
```diff
@@ -176,6 +187,7 @@ impl Chain {
 					genesis.header.nonce,
 					genesis.header.pow,
 				);
+				let _ = pipe::save_pmmr_metadata(&tip, &sumtrees, store.clone());
```
Comment (antiochp) [inline chain/src/chain.rs]: Probably needs a `?` here in case of failure?

**Pull request merged**